draft-holmberg-mmusic-ice-premature.xml

```diff
@@ -58,35 +58,33 @@
+        if connectivity can be established. If the test for a given pair fails,
+        it is marked accordingly, and if all pairs have failed, the overall
+        ICE process is considered to have failed.
       </t>
-      <t>
-        NOTE: Future specifications might describe how FQDNs that resolve into multiple connection-addresses
-        are used with ICE.
-      </t>
-      <t>
-        According to the procedures in <xref target="RFC8445"></xref>, if an endpoint discards the complete set
-        of initial candidates received from the remote peer, it will declare an ICE failure, and terminate
-        the session establishment, as the endpoint's list of remote candidates will be empty.
+      <t>During the process of connectivity checks, additional candidates may
+        be created as a result of successful inbound checks. Such candidates
+        are referred to as peer-reflexive candidates, and once discovered, will
+        be used to form new candidate pairs which will be tested like any other.
+        However, there is an inherent race condition here; if an endpoint
+        runs out of viable candidate pairs before learning about the
+        peer-reflexive candidates, it will prematurely fail the overall ICE
+        process. This race condition can occur in many common situations.
       </t>
       <t>
-        Once the endpoints have exchanged the initial set of candidates, and start performing connectivity
-        cheks, additional candidiates might be created as a result of the connectivity checks. Such candidates
-        are referred to as peer relexive candidates. However, the endpoint will most likely have declared
-        ICE failure before it is able to create such candidates.
+        This specification updates <xref target="RFC8445"></xref>, by simply
+        requiring that an endpoint wait a minimum amount of time before
+        declaring ICE failure, even if all candidate pairs have failed, or there
+        are no candidate pairs to check. This delay provides enough time for the
+        discovery of peer-reflexive candidates, which may eventually lead to the
+        success of the ICE process.
       </t>
       <t>
-        This specifcation updates <xref target="RFC8445"></xref>, by suggesting that an endpoint does not immediately
-        declare ICE failure in case the initial set of candidiates received from the remote peer are
-        discarded, but instead wait for a while in case it will receive connectivity checks that create one
-        or more peer reflexive candidates. Once the endpoint has created such candidates, it processes them
-        using the normal procedures defined in <xref target="RFC8445"></xref>.
+        One consequence of this change is that ICE will never fail fast, but
+        given that fast failure only occurred in specific scenarios, this will
+        have negligible real-world downside.
       </t>
```

Comment:
Instead of saying "will never fail fast" I would like to say "will not fail if there are peer-reflexive candidates that will form working candidate pairs", or something like that...

In addition, somewhere in the document we should clarify that the solution in the specification will not solve cases where BOTH endpoints indicate that all pairs fail.

Reply:
I was trying to say here that even in cases where ICE probably should fail immediately, e.g., cases where both sides have only nonroutable host candidates, it will not be able to. Will try to clarify.

Reply:
I moved this paragraph to later in the document where it's easier to explain.